System.Reflection.Emit: ParameterBuilder.SetConstant(null) throws for parameters of type Nullable<T>

Given a `ParameterBuilder` for a parameter of type `Nullable<T>`, a call to [`ParameterBuilder.SetConstant(null)`](https://github.com/mono/mono/blob/89ecd02722d1aa484dca476a800435639b544fd8/mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs#L95-L109 "see source code for `ParameterBuilder.SetConstant`") will throw an exception, claiming that `null` is not a supported constant type.

However, it should be perfectly possible to assign a null reference to such a parameter: The Roslyn C# compiler will turn this C# type definition:

```csharp
public interface IInterface
{
    void Method(int? arg = null);
}
```

into this IL:

```
.class interface public auto ansi abstract IInterface
{
    .method public hidebysig newslot abstract virtual instance void Method ([opt] valuetype [mscorlib]System.Nullable`1<int32> arg) cil managed 
    {
        .param [1] = nullref
    }
}
```

## Steps to reproduce

Compile and run the following program:

```csharp
using System;
using System.Reflection;
using System.Reflection.Emit;

class Program
{
    static void Main()
    {
        var assemblyBuilder = AppDomain.CurrentDomain.DefineDynamicAssembly(new AssemblyName("DynamicAssembly"), AssemblyBuilderAccess.RunAndSave);
        var moduleBuilder = assemblyBuilder.DefineDynamicModule("DynamicModule", "DynamicAssembly.dll");
        var typeBuilder = moduleBuilder.DefineType("IInterface", TypeAttributes.Abstract | TypeAttributes.Interface);
        var methodBuilder = typeBuilder.DefineMethod("Method", MethodAttributes.Public | MethodAttributes.Abstract | MethodAttributes.Virtual, typeof(void), new[] { typeof(int?) });
        var parameterBuilder = methodBuilder.DefineParameter(1, ParameterAttributes.Optional | ParameterAttributes.HasDefault, "arg");
        parameterBuilder.SetConstant(null);
    }
}
```

### Current behavior

`parameterBuilder.SetConstant(null)` causes the following run-time exception:

> System.ArgumentException: System.Nullable`1[System.Int32] is not a supported constant type.

### Expected behavior

The call to `parameterBuilder.SetConstant(null)` should succeed for parameters with type `Nullable<T>`, as it does on the CLR / on Windows.

## Platform on which I noticed this

[ ] macOS
[X] Bash on Ubuntu on Windows subsystem of Windows 10 (Ubuntu 16.04.4 LTS)
[ ] Windows

**Version used**:

Mono 5.10.1.47

## Stacktrace

```
Unhandled Exception:
System.ArgumentException: System.Nullable`1[System.Int32] is not a supported constant type.
  at System.Reflection.Emit.ParameterBuilder.SetConstant (System.Object defaultValue) [0x0008c] in <9bbab8f8a2a246e98480e70b0839fd67>:0
  at Program.Main () [0x0007a] in <aef196c188e64047a149a0675e76255d>:0
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: System.Nullable`1[System.Int32] is not a supported constant type.
  at System.Reflection.Emit.ParameterBuilder.SetConstant (System.Object defaultValue) [0x0008c] in <9bbab8f8a2a246e98480e70b0839fd67>:0
  at Program.Main () [0x0007a] in <aef196c188e64047a149a0675e76255d>:0
```

--- a/mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs
+++ b/mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs
@@ -96,12 +96,24 @@ namespace System.Reflection.Emit {
 		{
 			if (position > 0) {
 				Type t = methodb.GetParameterType (position - 1);
-				if (defaultValue != null && t != defaultValue.GetType ()) {
-					if(!t.IsEnum || t.UnderlyingSystemType != defaultValue.GetType ())
-						throw new ArgumentException ("Constant does not match the defined type.");
+				if (defaultValue == null) {
+					if (!(t.IsGenericType && t.GetGenericTypeDefinition () == typeof(Nullable<>)))
+						throw new ArgumentException("Null is not a valid constant value for this type");
+				} else {
+					if (t.IsEnum) {
+						Type underlyingType = Enum.GetUnderlyingType (t);
+						if (defaultValue.GetType () != underlyingType && defaultValue.GetType () != t) {
+							throw new ArgumentException ("Constant does not match the defined type.");
+						}
+					} else {
+						if (!t.IsAssignableFrom (defaultValue.GetType ()))
+							throw new ArgumentException ("Constant does not match the defined type.");
+					}
+
+					if (t.IsValueType && !t.IsPrimitive && !t.IsEnum && t != typeof (DateTime) &&
+					    !(t.IsGenericType && t.GetGenericTypeDefinition () == typeof (Nullable<>)))
+						throw new ArgumentException ("" + t + " is not a supported constant type.");
 				}
-				if (t.IsValueType && !t.IsPrimitive && !t.IsEnum && t != typeof (DateTime))
-					throw new ArgumentException ("" + t + " is not a supported constant type.");
 			}
 
 			def_value = defaultValue;

--- a/mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs
+++ b/mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs
@@ -95,25 +95,8 @@ namespace System.Reflection.Emit {
 		public virtual void SetConstant (object defaultValue)
 		{
 			if (position > 0) {
-				Type t = methodb.GetParameterType (position - 1);
-				if (defaultValue == null) {
-					if (!(t.IsGenericType && t.GetGenericTypeDefinition () == typeof(Nullable<>)))
-						throw new ArgumentException("Null is not a valid constant value for this type");
-				} else {
-					if (t.IsEnum) {
-						Type underlyingType = Enum.GetUnderlyingType (t);
-						if (defaultValue.GetType () != underlyingType && defaultValue.GetType () != t) {
-							throw new ArgumentException ("Constant does not match the defined type.");
-						}
-					} else {
-						if (!t.IsAssignableFrom (defaultValue.GetType ()))
-							throw new ArgumentException ("Constant does not match the defined type.");
-					}
-
-					if (t.IsValueType && !t.IsPrimitive && !t.IsEnum && t != typeof (DateTime) &&
-					    !(t.IsGenericType && t.GetGenericTypeDefinition () == typeof (Nullable<>)))
-						throw new ArgumentException ("" + t + " is not a supported constant type.");
-				}
+				TypeBuilder.SetConstantValue (methodb.GetParameterType (position - 1),
+							      defaultValue, ref defaultValue);
 			}
 
 			def_value = defaultValue;

--- a/mcs/class/corlib/System.Reflection.Emit/TypeBuilder.cs
+++ b/mcs/class/corlib/System.Reflection.Emit/TypeBuilder.cs
@@ -1903,6 +1903,125 @@ namespace System.Reflection.Emit
 		{
 			return base.IsAssignableFrom (typeInfo);
 		}
+
+		internal static bool SetConstantValue (Type destType, Object value, ref Object destValue)
+		{
+			// Mono: This is based on the CoreCLR
+			// TypeBuilder.SetConstantValue except it writes to an
+			// out argument instead of doing an icall, and it uses
+			// TypeCode instead of CorElementType (like
+			// MonoTypeEnum) which we don't have in our corlib and
+			// our member fields are different.
+
+			// This is a helper function that is used by ParameterBuilder, PropertyBuilder,
+			// and FieldBuilder to validate a default value and save it in the meta-data.
+
+			if (value != null) {
+				Type type = value.GetType ();
+
+				// We should allow setting a constant value on a ByRef parameter
+				if (destType.IsByRef)
+					destType = destType.GetElementType ();
+
+				// Convert nullable types to their underlying type.
+				// This is necessary for nullable enum types to pass the IsEnum check that's coming next.
+				destType = Nullable.GetUnderlyingType (destType) ?? destType;
+
+				if (destType.IsEnum)
+				{
+					//                                   |  UnderlyingSystemType     |  Enum.GetUnderlyingType() |  IsEnum
+					// ----------------------------------|---------------------------|---------------------------|---------
+					// runtime Enum Type                 |  self                     |  underlying type of enum  |  TRUE
+					// EnumBuilder                       |  underlying type of enum  |  underlying type of enum* |  TRUE
+					// TypeBuilder of enum types**       |  underlying type of enum  |  Exception                |  TRUE
+					// TypeBuilder of enum types (baked) |  runtime enum type        |  Exception                |  TRUE
+
+					//  *: the behavior of Enum.GetUnderlyingType(EnumBuilder) might change in the future
+					//     so let's not depend on it.
+					// **: created with System.Enum as the parent type.
+
+					// The above behaviors might not be the most consistent but we have to live with them.
+
+					Type underlyingType;
+					EnumBuilder enumBldr;
+					TypeBuilder typeBldr;
+					if ((enumBldr = destType as EnumBuilder) != null) {
+						underlyingType = enumBldr.GetEnumUnderlyingType ();
+
+						// The constant value supplied should match either the baked enum type or its underlying type
+						// we don't need to compare it with the EnumBuilder itself because you can never have an object of that type
+						if (!((enumBldr.GetTypeBuilder ().is_created && type == enumBldr.GetTypeBuilder ().created) ||
+						      type == underlyingType))
+							throw_argument_ConstantDoesntMatch ();
+					} else if ((typeBldr = destType as TypeBuilder) != null) {
+						underlyingType = typeBldr.underlying_type;
+
+						// The constant value supplied should match either the baked enum type or its underlying type
+						// typeBldr.m_enumUnderlyingType is null if the user hasn't created a "value__" field on the enum
+						if (underlyingType == null || (type != typeBldr.UnderlyingSystemType && type != underlyingType))
+							throw_argument_ConstantDoesntMatch ();
+					} else {
+						// must be a runtime Enum Type
+
+						// Debug.Assert(destType is RuntimeType, "destType is not a runtime type, an EnumBuilder, or a TypeBuilder.");
+
+						underlyingType = Enum.GetUnderlyingType (destType);
+
+						// The constant value supplied should match either the enum itself or its underlying type
+						if (type != destType && type != underlyingType)
+							throw_argument_ConstantDoesntMatch ();
+					}
+
+					type = underlyingType;
+				} else {
+					// Note that it is non CLS compliant if destType != type. But RefEmit never guarantees CLS-Compliance.
+					if (!destType.IsAssignableFrom (type))
+						throw_argument_ConstantDoesntMatch ();
+				}
+
+				TypeCode corType = Type.GetTypeCode (type);
+
+				switch (corType)
+				{
+					case TypeCode.Byte:
+					case TypeCode.SByte:
+					case TypeCode.Boolean:
+					case TypeCode.Int16:
+					case TypeCode.UInt16:
+					case TypeCode.Char:
+					case TypeCode.Int32:
+					case TypeCode.UInt32:
+					case TypeCode.Single:
+					case TypeCode.Int64:
+					case TypeCode.UInt64:
+					case TypeCode.Double:
+						destValue = value;
+						return true;
+					case TypeCode.String:
+						destValue = value;
+						return true;
+					case TypeCode.DateTime:
+						//date is a I8 representation
+						long ticks = ((DateTime)value).Ticks;
+						destValue = ticks;
+						return true;
+					default:
+						throw new ArgumentException(type.ToString() + " is not a supported constant type.");
+				}
+			} else {
+				// A null default value in metadata is permissible even for non-nullable value types.
+				// (See ECMA-335 II.15.4.1.4 "The .param directive" and II.22.9 "Constant" for details.)
+				// This is how the Roslyn compilers generally encode `default(TValueType)` default values.
+
+				destValue = null;
+				return true;
+			}
+		}		
+
+		private static void throw_argument_ConstantDoesntMatch ()
+		{
+			throw new ArgumentException("Constant does not match the defined type.");
+		}
 	}
 }
 #endif

--- a/mcs/class/corlib/Test/System.Reflection.Emit/ParameterBuilderTest.cs
+++ b/mcs/class/corlib/Test/System.Reflection.Emit/ParameterBuilderTest.cs
@@ -47,5 +47,48 @@ namespace MonoTests.System.Reflection.Emit
 			var builder = hello.DefineParameter(1, ParameterAttributes.In, "failParam");
 			builder.SetConstant(1);
 		}
+
+		public enum E : byte {
+			E1, E2
+		}
+
+		[Test]
+		public void SetConstantNullable ()
+		{
+			// SetConstant  for a Nullable<X> parameter for various X.
+			var aName = new AssemblyName ("TestSetConstantNullable");
+			var ab = AppDomain.CurrentDomain.DefineDynamicAssembly (aName, AssemblyBuilderAccess.Run);
+
+			var mb = ab.DefineDynamicModule (aName.Name);
+
+			var tb = mb.DefineType ("TestItf", TypeAttributes.Abstract | TypeAttributes.Interface);
+
+			var paramTypes = new[] { typeof (int?), typeof (int?),
+						 typeof (E?), typeof (E?) };
+
+			var methb = tb.DefineMethod ("Method", MethodAttributes.Public | MethodAttributes.Abstract | MethodAttributes.Virtual, typeof (void), paramTypes);
+			var pb = new ParameterBuilder [paramTypes.Length];
+			for (int idx = 0, pnum = 1; idx < paramTypes.Length; idx++, pnum++) {
+				pb [idx] = methb.DefineParameter (pnum, ParameterAttributes.Optional | ParameterAttributes.HasDefault, "arg" + idx);
+			}
+
+			pb [0].SetConstant (null);
+			pb [1].SetConstant (42);
+			pb [2].SetConstant (null);
+			pb [3].SetConstant (E.E2);
+
+			var t = tb.CreateType ();
+
+			var mi = t.GetMethod ("Method");
+
+			Assert.IsNotNull (mi);
+
+			var ps = mi.GetParameters ();
+
+			Assert.AreEqual (null, ps[0].DefaultValue);
+			Assert.AreEqual (42, ps[1].DefaultValue);
+			Assert.AreEqual (null, ps[2].DefaultValue);
+			Assert.AreEqual ((byte)E.E2, ps[3].DefaultValue);
+		}
 	}
 }

